core/includes/config.inc

```diff
@@ -513,6 +632,39 @@ class Config {
   }
 
   /**
+   * Merge new defaults with this configuration object and update the default
+   * hash signature.  This method will not overwrite existing configuration
+   * settings.  It only adds settings that are missing.
+   *
+   * @param array $data
+   *   The new default configuration data.
+   *
+   * @return Config
+   *   The configuration object.
+   */
+  public function mergeDefaultData(array $data) {
+    $this->replaceData($this->data + $data);
+    $this->setDefaultHash($data);
+    // A load would destroy the data just set (for example on import).
+    $this->isLoaded = TRUE;
+    return $this;
+  }
+
+  /**
+   * Computes the hash signature of the given config data or the current config
+   * data and save it in the _default_hashes section.
+   *
+   * @param array $data
+   *   The data to hash.  Uses the current configuration data set if NULL.
+   */
+  public function setDefaultHash($data=NULL) {
+    if ($data === NULL) {
+      $data = $this->data;
+    }
+    $algo = self::DEFAULT_HASH_ALGORITHM;
+    $this->data['_default_hashes'][$algo] = Config::Hash($data, $algo);
```

Comment:
The algorithm used for a hash in this case isn't really relevant. This hash isn't being used for encryption, it's only used to check if a file has changed. You're right that hashes have been changed over time for attempting to secure information (passwords, credit card data, etc), but that's not what we're doing here. We just want to know when a file is changed. If we truly needed to change the hash at some point, we could add a second key, e.g. `_default_hash_algo` that stored the hash type, defaulting to the original hash method (sha1). There'd be no reason to store *multiple* hashes at the same time.

Even then if we change the hash, we could bulk-update all the existing config files at the same time via update hook, though we'd probably only do so at a major release (1.x to 2.x). There already have been lots of situations where we need to update existing config files by update hook, not for setting defaults but for moving or extending options. So we're not necessarily tied to what we choose here anyway.